### PR TITLE
Added bottom tab bar

### PR DESCRIPTION
**Pull request by hernan-clich**
## Summary

Added the bottom tab navigation which includes the `Library` screen as well as an empty dummy screen.

#### Galaxy S (or similar)

tabbar_galaxyS

#### Pixel 2/3 (or similar)

tabbar_pixel3

#### Pixel 3XL (or similar)

tabbar_pixel3xl

## Trello Card

https://trello.com/c/GyNaqwUa/18-n-8-tab-bar

### CODE REVIEW

File: wbooks/src/app/components/AppNavigator/index.tsx [hunk 1]
```diff
@@ -30,7 +31,11 @@ const AppNavigator = () => {
           textTransform: 'uppercase'
         }
       }}>
-      <Stack.Screen name={SCREENS.LIBRARY} component={Library} options={{ title: 'Library' }} />
+      <Stack.Screen
+        name={SCREENS.HOME}
+        component={TabNavigator}
+        options={({ route }) => ({ headerTitle: getHeaderTitle(route) })}
+      />
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/index.tsx]: Wrong implementation, please revert changes and re-do this. Check again in the docs

File: wbooks/src/app/components/AppNavigator/utils/getHeaderTitle.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { getFocusedRouteNameFromRoute, RouteProp } from '@react-navigation/native';
+import { HomeType } from '@constants/screens';
+import { RootStackParamList } from '@interfaces/navigatorParamLists';
+
+function getHeaderTitle(route: RouteProp<RootStackParamList, HomeType>): string {
+  const routeName = getFocusedRouteNameFromRoute(route) ?? 'Library';
+
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/utils/getHeaderTitle.ts]: Unnecessary

File: wbooks/src/app/components/TabIcon/index.tsx [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import { Image, ImageProps } from 'react-native';
+
+interface Props {
+  active: ImageProps;
+  inactive: ImageProps;
+  focused: boolean;
+}
+
+const TabIcon = ({ active, inactive, focused }: Props) => {
+  return <Image resizeMode="contain" source={focused ? active : inactive} />;
```
Comment (sfernandez11) [inline wbooks/src/app/components/TabIcon/index.tsx]: You must change this implementation. Use `useRoute` hook to know the current route and then:
```
source={TAB_ICONS[route.name]}
```
and
```
TAB_ICONS = {
[ROUTE]: { active: ..., inactive: ... } 
}
```
or something like that

File: wbooks/src/app/components/TabNavigator/index.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  return (
+    <Tab.Navigator tabBarOptions={{ activeTintColor: colors.cerulean, inactiveTintColor: colors.dustyGray }}>
+      <Tab.Screen
+        name={SCREENS.LIBRARY}
+        component={Library}
+        options={{
+          title: 'Library',
+          tabBarIcon: ({ focused }) => (
+            <TabIcon active={libraryIconActive} inactive={libraryIcon} focused={focused} />
+          )
```
Comment (sfernandez11) [inline wbooks/src/app/components/TabNavigator/index.tsx]: Move this to `Tab.Navigator` config =>
```
<Tab.Navigator ... screenOptions={{ 
  tabBarIcon: ({ focused }) => (
    <TabIcon focused={focused} />
)}}
```

File: wbooks/src/app/components/TabNavigator/index.tsx [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import Library from '@screens/Library';
+
+const Tab = createBottomTabNavigator<bottomTabParamList>();
+
+const TabNavigator = () => {
+  return (
+    <Tab.Navigator tabBarOptions={{ activeTintColor: colors.cerulean, inactiveTintColor: colors.dustyGray }}>
+      <Tab.Screen
+        name={SCREENS.LIBRARY}
+        component={Library}
```
Comment (sfernandez11) [inline wbooks/src/app/components/TabNavigator/index.tsx]: Here must be the `Library` stack, not only the screen

File: wbooks/src/app/components/TabNavigator/index.tsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        }}
+      />
+      <Tab.Screen
+        name={SCREENS.WISHLIST}
+        component={EmptyView}
+        options={{
+          title: 'Wishlist',
+          tabBarIcon: ({ focused }) => (
+            <TabIcon active={wishlistIconActive} inactive={wishlistIcon} focused={focused} />
+          )
```
Comment (sfernandez11) [inline wbooks/src/app/components/TabNavigator/index.tsx]: Delete

File: wbooks/src/app/components/TabNavigator/index.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import TabIcon from '@components/TabIcon';
+import colors from '@constants/colors';
+import SCREENS from '@constants/screens';
+import { bottomTabParamList } from '@interfaces/navigatorParamLists';
+import EmptyView from '@screens/EmptyView';
+import Library from '@screens/Library';
+
+const Tab = createBottomTabNavigator<bottomTabParamList>();
+
+const TabNavigator = () => {
```
Comment (sfernandez11) [inline wbooks/src/app/components/TabNavigator/index.tsx]: Move this inside `AppNavigator` file

File: wbooks/src/app/screens/BookDetails/index.tsx [hunk 8]
```diff
@@ -1,8 +1,8 @@
 import React from 'react';
-import { View } from 'react-native';
+import EmptyView from '@screens/EmptyView';
 
 function BookDetails() {
-  return <View />;
+  return <EmptyView />;
```
Comment (sfernandez11) [inline wbooks/src/app/screens/BookDetails/index.tsx]: No, conceptually wrong. You must use components, not screens. Also, unnecessary

File: wbooks/src/app/screens/Library/index.tsx [hunk 9]
```diff
@@ -3,15 +3,15 @@ import { FlatList, ListRenderItem, SafeAreaView } from 'react-native';
...
 import { Book } from '@interfaces/books';
-import { RootStackParamList } from '@interfaces/rootStackParamList';
+import { RootStackParamList } from '@interfaces/navigatorParamLists';
 import { trimLineBreaks } from '@utils/stringUtils';
 
 import styles from './styles';
 
 interface Props {
-  navigation: StackNavigationProp<RootStackParamList, LibraryType>;
+  navigation: StackNavigationProp<RootStackParamList, HomeType>;
```
Comment (sfernandez11) [inline wbooks/src/app/screens/Library/index.tsx]: Revert. this must be `Library`, not `Home`

File: wbooks/src/constants/screens.ts [hunk 10]
```diff
@@ -1,14 +1,20 @@
...
+export type LibraryType = 'Library';
+export type WishListType = 'Wishlist';
 
 interface Screens {
-  LIBRARY: LibraryType;
   BOOK_DETAILS: BookDetailType;
+  HOME: HomeType;
+  LIBRARY: LibraryType;
+  WISHLIST: WishListType;
 }
```
Comment (sfernandez11) [inline wbooks/src/constants/screens.ts]: Interfaces must be inside `interfaces` folder

File: wbooks/src/interfaces/navigatorParamLists.ts [hunk 11]
```diff
@@ -1,4 +1,9 @@
 export type RootStackParamList = {
-  Library: undefined;
   BookDetails: undefined;
+  Home: undefined;
+};
+
+export type bottomTabParamList = {
```
Comment (sfernandez11) [inline wbooks/src/interfaces/navigatorParamLists.ts]: We use PascalCase for `screens`, `components` and `interfaces`

**Review by sfernandez11 — CHANGES_REQUESTED** [review]
Please check the Tab implementation, some comments...

File: wbooks/src/app/components/AppNavigator/index.tsx (hunk 1 above)
Comment (hernan-clich) [reply wbooks/src/app/components/AppNavigator/index.tsx]: I obtained this method from the [docs](https://reactnavigation.org/docs/screen-options-resolution/#setting-parent-screen-options-based-on-child-navigators-state) and is their recommended method.

File: wbooks/src/app/components/TabNavigator/index.tsx (hunk 4 above)
Comment (hernan-clich) [reply wbooks/src/app/components/TabNavigator/index.tsx]: Refactored

File: wbooks/src/app/components/TabNavigator/index.tsx (hunk 6 above)
Comment (hernan-clich) [reply wbooks/src/app/components/TabNavigator/index.tsx]: Deleted

File: wbooks/src/app/components/TabNavigator/index.tsx (hunk 7 above)
Comment (hernan-clich) [reply wbooks/src/app/components/TabNavigator/index.tsx]: Do you mean having both components declared on the same file?

File: wbooks/src/app/components/TabIcon/index.tsx (hunk 3 above)
Comment (hernan-clich) [reply wbooks/src/app/components/TabIcon/index.tsx]: Refactored

File: wbooks/src/app/components/TabNavigator/index.tsx (hunk 5 above)
Comment (hernan-clich) [reply wbooks/src/app/components/TabNavigator/index.tsx]: Created a new stack navigator to display the Library as its only route

File: wbooks/src/app/components/AppNavigator/index.tsx [hunk 12]
```diff
@@ -30,10 +58,18 @@ const AppNavigator = () => {
           textTransform: 'uppercase'
         }
       }}>
-      <Stack.Screen name={SCREENS.LIBRARY} component={Library} options={{ title: 'Library' }} />
-      <Stack.Screen name={SCREENS.BOOK_DETAILS} component={BookDetails} options={{ title: 'Book Detail' }} />
+      <Stack.Screen
+        name={SCREENS.LIBRARY}
+        component={TabNavigator}
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/index.tsx]: This implementation is still wrong, please re-read the docs

File: wbooks/src/app/components/AppNavigator/index.tsx (hunk 1 above)
Comment (sfernandez11) [reply wbooks/src/app/components/AppNavigator/index.tsx]: I talked about the stack/tab implementation, not the `getHeaderTitle`. Btw, I think this method is not necessary either

File: wbooks/src/app/components/TabIcon/index.tsx [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+
+import TAB_ICONS from './constants';
+
+interface Props {
+  focused: boolean;
+}
+
+function TabIcon({ focused }: Props) {
+  const route = useRoute();
+  const source = focused ? TAB_ICONS[route.name].inactive : TAB_ICONS[route.name].active;
```
Comment (sfernandez11) [inline wbooks/src/app/components/TabIcon/index.tsx]: `const source = TAB_ICONS[route.name][focused ? 'active' : 'inactive'];`

File: wbooks/src/app/components/AppNavigator/index.tsx [hunk 14]
```diff
@@ -1,19 +1,47 @@
...
 
 const Stack = createStackNavigator<RootStackParamList>();
+const Tab = createBottomTabNavigator<BottomTabParamList>();
 
-const AppNavigator = () => {
+function LibraryStack() {
+  return (
+    <Stack.Navigator headerMode="none">
+      <Stack.Screen name={SCREENS.LIBRARY} component={Library} options={{ title: SCREEN_TITLES.LIBRARY }} />
+    </Stack.Navigator>
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/index.tsx]: Wrong

**Review by sfernandez11 — CHANGES_REQUESTED** [review]
Please fix the TabBar implementation. Check the first example of the [docs](https://reactnavigation.org/docs/screen-options-resolution) because this is the correct approach

**Comment by hernan-clich** [comment]
I followed strictly the example from the docs that you left in one of your last comments and set the Tab navigator as the main navigation of the app.

File: wbooks/src/app/components/AppNavigator/config/customHeader.tsx [hunk 15]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import { Image } from 'react-native';
+import backButton from '@assets/ic_back.png';
+import headerBackground from '@assets/bc_nav_bar.png';
+import colors from '@constants/colors';
+
+import styles from '../styles';
+
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/config/customHeader.tsx]: Wrong folder structure. Why did you create a `config` folder inside `AppNavigator`  component?

Move `wbooks/src/app/components/AppNavigator/config/customHeader.tsx` => `wbooks/src/app/components/AppNavigator/constants.tsx` or `.../config.tsx`

We only use folders for `assets`, `components` or `screens` inside other `components`/`screens`

**Review by sfernandez11 — COMMENTED** [review]
Last comment

File: wbooks/src/app/components/AppNavigator/config/customHeader.tsx (hunk 15 above)
Comment (hernan-clich) [reply wbooks/src/app/components/AppNavigator/config/customHeader.tsx]: Changed from `AppNavigator/config/customHeader.tsx` to `AppNavigator/config.tsx`

**Review by sfernandez11 — APPROVED** [review]
🥇

**Pull request merged**